File: 1-TwoSum/1-TwoSum.py
# Last updated: 6/27/2026, 10:37:11 PM
1class Solution:
2    # def twoSum(self, nums: List[int], target: int) -> List[int]:
3    #     # brute force
4    #     n = len(nums)
5    #     for i in range(n):
6    #         for j in range(i + 1, n):
7    #             if nums[i] + nums[j] == target:
8    #                 return [i, j]
9        
10    #     return [-1, -1]
11    
12    def twoSum(self, nums: List[int], target: int) -> List[int]:
13        # hashmap
14        # if scan once, for each element, check if visited "target - curr" before
15        # hashmap key = curr, value = index
16        seen = {}
17        n = len(nums)
18        for i in range(n):
19            new_target = target - nums[i]
20            if new_target in seen:
21                return [seen[new_target], i]
22            seen[nums[i]] = i
23            # print(seen)
24        return [-1, -1]
25

File: 1-TwoSum/1-TwoSum.java
// Last updated: 8/16/2026, 9:54:02 PM

class Solution {
    public int[] twoSum(int[] nums, int target) {
        /*
        // 1. Brute Force
        for (int i = 0; i < nums.length; i ++) {
            for (int j = i + 1; j < nums.length; j ++) {
                if (nums[i] + nums[j] == target) {
                    return new int[]{i, j};
                }
            }
        }
        return null;
        
        // 2. HashMap - two pass
        Map<Integer, Integer> map = new HashMap<>();
        int n = nums.length;
        for (int i = 0; i < n; i ++) {
            map.put(nums[i], i);
        }
        for (int i = 0; i < n; i ++) {
            int newTarget = target - nums[i];
            if (map.containsKey(newTarget) && map.get(newTarget) != i) {
                return new int[]{i, map.get(newTarget)};
            }
        }
        return null;
        */

        // 3.HashMap - One pass
        Map<Integer, Integer> map = new HashMap<>();
        int n = nums.length;
        for (int i = 0; i < n; i ++) {
            int newTarget = target - nums[i];
            if (map.containsKey(newTarget)) {
                return new int[]{i, map.get(newTarget)};
            }
            map.put(nums[i], i);
        }
        return null;
    }
}

